### PR TITLE
Align addition of a synthesized override of object.Equals(object? obj) in records with the latest design.

### PR DESCRIPTION
**Pull request by AlekseyTs**
Related to #45296.

From specification:
The record type includes a synthesized override of object.Equals(object? obj). It is an error if the override is declared explicitly. The synthesized override returns Equals(other as R) where R is the record type.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@jcouv, @agocke, @cston , @RikkiGibson, @dotnet/roslyn-compiler Please review.

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordObjEquals.cs [hunk 1]
```diff
@@ -4,118 +4,119 @@
...
-        public override NamedTypeSymbol ContainingType { get; }
-
-        public override ImmutableArray<ParameterSymbol> Parameters { get; }
-
-        public SynthesizedRecordObjEquals(NamedTypeSymbol containingType, MethodSymbol typedRecordEquals, int memberOffset)
+        public SynthesizedRecordObjEquals(SourceMemberContainerTypeSymbol containingType, MethodSymbol typedRecordEquals, int memberOffset, DiagnosticBag diagnostics)
+            : base(containingType, "Equals", containingType.Locations[0], (CSharpSyntaxNode)containingType.SyntaxReferences[0].GetSyntax(), MethodKind.Ordinary,
+                   isIterator: false, isExtensionMethod: false, isPartial: false, hasBody: true, diagnostics)
         {
             var compilation = containingType.DeclaringCompilation;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordObjEquals.cs:25]: >var compilation = containingType.DeclaringCompilation; [](start = 12, length = 54)

This line is no longer needed.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 2]
```diff
@@ -3010,6 +3010,13 @@ private void AddSynthesizedRecordMembersIfNecessary(MembersAndInitializersBuilde
             otherEqualsMethods.Free();
             memberSignatures.Free();
 
+
+            // We put synthesized record members first so that errors about conflicts show up on user-defined members rather than all
+            // go to the record declaration
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:3015]: >go [](start = 15, length = 2)

Small nit: go -> going

**Comment by AlekseyTs** [comment]
@jcouv, @agocke, @cston, @dotnet/roslyn-compiler Please review, need a second sign-off.

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordObjectMethod.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        {
+        }
+
+        protected sealed override void MethodChecks(DiagnosticBag diagnostics)
+        {
+            base.MethodChecks(diagnostics);
+
+            var overridden = OverriddenMethod?.OriginalDefinition;
+
+            if (overridden is null || (overridden is SynthesizedRecordObjEquals && overridden.DeclaringCompilation == DeclaringCompilation))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordObjectMethod.cs:25]: >SynthesizedRecordObjEquals [](start = 53, length = 26)

This should be adjusted to handle a general case, but this PR doesn't have any. Fixed in the change that builds on top of this PR.

**Comment by AlekseyTs** [comment]
@jcouv, @agocke, @cston, @dotnet/roslyn-compiler Please review, need a second sign-off.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 4]
```diff
@@ -2965,8 +2965,8 @@ private void AddSynthesizedRecordMembersIfNecessary(MembersAndInitializersBuilde
             ParameterListSyntax? paramList = builder.RecordDeclarationWithParameters?.ParameterList;
 
             var memberSignatures = s_duplicateMemberSignatureDictionary.Allocate();
-            var members = builder.NonTypeNonIndexerMembers;
-            foreach (var member in members)
+            var members = ArrayBuilder<Symbol>.GetInstance(builder.NonTypeNonIndexerMembers.Count + 1);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:2968]: >members [](start = 16, length = 7)

Consider moving to line 3015.

**Pull request merged**